Game/src/Game.cpp

```diff
@@ -63,5 +68,52 @@ void Game::Init()
   menu_->AddSubMenu({ "end game", endGame });
 }
 
+void Game::Action()
+{
+  std::lock_guard<std::mutex> _(mutex_);
+
+  if(glfwGetKey(window_, GLFW_KEY_LEFT) == GLFW_PRESS)
+  {
+    // Moving right
+    std::cout << "moving right" << std::endl;
+  }
+  if(glfwGetKey(window_, GLFW_KEY_RIGHT) == GLFW_PRESS)
+  {
+    // moving left
+    std::cout << "moving left" << std::endl;
+  }
+  if(glfwGetKey(window_, GLFW_KEY_SPACE) == GLFW_PRESS)
+  {
+    // Jump
+    std::cout << "jump" << std::endl;
+  }
+}
+
+void Game::Render()
+{
+  /* Render here */
+  glClear(GL_COLOR_BUFFER_BIT);
+
+  viz_->OnRender();
+
+  /* Swap front and back buffers */
+  glfwSwapBuffers(window_);
+
+  /* Poll for and process events */
+  glfwPollEvents();
+}
+
+void Game::Mainloop()
+{
+  while (!glfwWindowShouldClose(window_))
+  {
+    std::thread threadAction(&Game::Action, this);
+    if(threadAction.joinable())
+      threadAction.join();
```

Comment:
These operations also have a cost! Have you tried to use profiler to search for fundamental differences?